icu4c/source/test/fuzzer/calendar_fuzzer.cpp

```diff
@@ -0,0 +1,125 @@
+// © 2023 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+// Fuzzer for ICU Calendar.
+
+#include <cstring>
+
+#include "fuzzer_utils.h"
+
+#include "unicode/calendar.h"
+#include "unicode/localebuilder.h"
+#include "unicode/locid.h"
+
+icu::TimeZone* CreateRandomTimeZone(uint16_t rnd) {
+    icu::Locale und("und");
+    UErrorCode status = U_ZERO_ERROR;
+    std::unique_ptr<icu::StringEnumeration> enumeration(
+        icu::TimeZone::createEnumeration(status));
+    if (U_SUCCESS(status)) {
+        int32_t count = enumeration->count(status);
+        if (U_SUCCESS(status)) {
+            int32_t i = rnd % count;
+            const icu::UnicodeString* id = nullptr;
+            do {
+              id = enumeration->snext(status);
+            } while (U_SUCCESS(status) && --i > 0);
+            if (U_SUCCESS(status)) {
+                return icu::TimeZone::createTimeZone(*id);
+            }
+        }
+    }
+    return icu::TimeZone::getGMT()->clone();
+}
+const char* GetRandomCalendarType(uint16_t rnd) {
+    icu::Locale und("und");
+    UErrorCode status = U_ZERO_ERROR;
+    std::unique_ptr<icu::StringEnumeration> enumeration(
+        icu::Calendar::getKeywordValuesForLocale("calendar", und, false, status));
+    const char* type = "";
+    if (U_SUCCESS(status)) {
+        int32_t count = enumeration->count(status);
+        if (U_SUCCESS(status)) {
+            int32_t i = rnd % count;
+            do {
+              type = enumeration->next(nullptr, status);
+            } while (U_SUCCESS(status) && --i > 0);
+        }
+    }
+    type = uloc_toUnicodeLocaleType("ca", type);
+    return type;
+}
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+    if (size < 2) return 0;
+    uint16_t rnd16;
+    std::memcpy(&rnd16, data, sizeof(rnd16));
+    data += 2;
+    size -= 2;
+    icu::Locale locale = GetRandomLocale(rnd16);
+    const char* type = GetRandomCalendarType(rnd16);
+    std::unique_ptr<icu::TimeZone> timeZone(CreateRandomTimeZone(rnd16));
+    UErrorCode status = U_ZERO_ERROR;
+    icu::LocaleBuilder bld;
+    bld.setLocale(locale);
+    bld.setUnicodeLocaleKeyword("ca", type);
+    locale = bld.build(status);
+    if (U_FAILURE(status)) return 0;
+    std::unique_ptr<icu::Calendar> cal(
+        icu::Calendar::createInstance(*timeZone, locale, status));
+    if (U_FAILURE(status)) return 0;
+    cal->clear();
+
+    while (size > 1 + sizeof(double)) {
+        uint8_t rnd8 = *data++;
+        uint32_t rnd32;
+        double date;
+        std::memcpy(&date, data, sizeof(date));
+        std::memcpy(&rnd32, data, sizeof(rnd32));
+        UCalendarDateFields field = static_cast<UCalendarDateFields>(rnd32 % UCAL_FIELD_COUNT);
+        int32_t srnd32 = (rnd32 % 2 == 0 ? -1 : 1) * rnd32;
```

Comment:
If I'm reading this correctly, it's saying that `srnd32` is the same as `rnd32` if `rnd32` is odd, and the negative of `rnd32` if `rnd32` is even.  So you're trying the sign to the ones bit.  I'm still not quite sure how this is being used, but that also seems to tie these two properties together in ways that won't give you true randomness.

In fact, I see `field` is derived from `rnd32` as well, so you're restricting yourself to certain combinations of field and increment value in the tests below.  Again, maybe this is okay, but it's not completely random.

Reply:
ok, I address all these now.